### PR TITLE
PReLu ONNX import

### PR DESCRIPTION
**Pull request by Arjun31415**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs
#1714 

### Changes

Added PReLu ONNX -> burn operator

### Testing

Tested as specified in the Contributor book

### CODE REVIEW

**Comment by Arjun31415** [comment]
The rust code produced - 
````rs
// Generated from ONNX "./onnx-tests/tests/prelu/prelu.onnx" by burn-import
use burn::nn::prelu::PRelu;
use burn::nn::prelu::PReluConfig;
use burn::nn::PRelu;
use burn::record::FullPrecisionSettings;
use burn::record::Recorder;
use burn::{
    module::Module,
    tensor::{backend::Backend, Tensor},
};


#[derive(Module, Debug)]
pub struct Model<B: Backend> {
    prelu1: PRelu<B>,
    phantom: core::marker::PhantomData<B>,
    device: burn::module::Ignored<B::Device>,
}


impl<B: Backend> Default for Model<B> {
    fn default() -> Self {
        Self::from_file("./out/prelu", &Default::default())
    }
}

impl<B: Backend> Model<B> {
    pub fn from_file(file: &str, device: &B::Device) -> Self {
        let record = burn::record::PrettyJsonFileRecorder::<FullPrecisionSettings>::new()
            .load(file.into(), device)
            .expect("Record file to exist.");
        Self::new(device).load_record(record)
    }
}

impl<B: Backend> Model<B> {
    #[allow(unused_variables)]
    pub fn new(device: &B::Device) -> Self {
        let prelu1 = PReluConfig::new(0, 0.009999999776482582).init(device);
        Self {
            prelu1,
            phantom: core::marker::PhantomData,
            device: burn::module::Ignored(device.clone()),
        }
    }

    #[allow(clippy::let_and_return, clippy::approx_constant)]
    pub fn forward(&self, input1: Tensor<B, 2>) -> Tensor<B, 0> {
        let prelu1_out1 = self.prelu1.forward(input1);
        prelu1_out1
    }
}
```
I think the `PReluConfig` initialization is wrong . 
```rs
        let prelu1 = PReluConfig::new(0, 0.009999999776482582).init(device);
```
the number of parameters should not be 0 and initial alpha should be 0.01 exactly and not this approximate value

File: crates/burn-import/src/burn/node/prelu.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        let output = &self.output.name;
+        let field = &self.field.name;
+
+        quote! {
+            let #output = self.#field.forward(#input);
+        }
+    }
+    fn register_imports(&self, imports: &mut BurnImports) {
+        imports.register("burn::nn::PRelu");
+        imports.register("burn::nn::prelu::PRelu");
```
Comment (antimora) [inline crates/burn-import/src/burn/node/prelu.rs:93]: PRelu seems to be imported twice here.

File: crates/burn-import/src/onnx/op_configuration.rs [hunk 2]
```diff
@@ -120,6 +120,21 @@ pub fn max_pool2d_config(curr: &Node) -> MaxPool2dConfig {
...
+pub fn prelu_config(curr: &Node) -> PReluConfig {
+    let mut alpha = 0.01;
+    let mut num_parameters = 0;
+    for (key, value) in curr.attrs.iter() {
+        match key.as_str() {
+            "alpha" => alpha = value.clone().into_f32(),
+            "num_parameters" => num_parameters = value.clone().into_i32(),
+            _ => {}
+        }
+    }
```
Comment (antimora) [inline crates/burn-import/src/onnx/op_configuration.rs:132]: ONNX Spec does not have alpha nor num_parameters, so they should be removed: https://onnx.ai/onnx/operators/onnx__PRelu.html

I recommend you would remove prelu_config function and just use PReluConfig::new() with it's defaults. You do not need to store or pass it.

config is meant to extract attributes (such as [Conv attributes](https://onnx.ai/onnx/operators/onnx__Conv.html#attributes))  as Module configuration.

You can also see the ONNX example you checked in does not contain alpha and num_params:

image

File: crates/burn-import/src/onnx/to_burn.rs [hunk 3]
```diff
@@ -695,6 +697,14 @@ impl OnnxGraph {
         MaxPool2dNode::new(name, input, output, config)
     }
 
+    fn prelu_conversion<PS: PrecisionSettings>(node: Node) -> PReluNode<PS> {
+        let input = node.inputs.first().unwrap().to_tensor_type();
+        let output = node.outputs.first().unwrap().to_tensor_type();
+        let weight = extract_data_serialize::<PS::FloatElem>(1, &node).unwrap();
+        let config = prelu_config(&node);
```
Comment (antimora) [inline crates/burn-import/src/onnx/to_burn.rs:704]: I would remove this too per my above comment.

**Review by antimora — CHANGES_REQUESTED** [review]
Great start but we are missing two types of unit tests (under onnx-tests and in crates/burn-import/src/burn/node/prelu.rs). Also Prelu ONNX spec does not have alpha and numb_parameters, so there is no need to extract configuration.

**Comment by antimora** [comment]
FYI, to run ONNX test:

```shell
[burn-import]% cd onnx-tests
[onnx-tests]% cargo test
```

to run code gen tests:
```sh
[burn-import]% cargo test
```

**Comment by Arjun31415** [comment]
I also get this warning 
```log
DEBUG burn_import::onnx::from_onnx: Number of initializers: 1    
DEBUG burn_import::onnx::from_onnx: Number of outputs: 1    
DEBUG burn_import::onnx::proto_conversion: Converting ONNX node with type "PRelu"    
DEBUG burn_import::onnx::from_onnx: renaming node "/relu1/PRelu"    
 WARN burn_import::onnx::dim_inference: Must implement dimension inference for PRelu    
DEBUG burn_import::onnx::from_onnx: checking inputs for node "prelu1"    
DEBUG burn_import::onnx::from_onnx: 
```

**Comment by antimora** [comment]
> I also get this warning
> 
> ```
> DEBUG burn_import::onnx::from_onnx: Number of initializers: 1    
> DEBUG burn_import::onnx::from_onnx: Number of outputs: 1    
> DEBUG burn_import::onnx::proto_conversion: Converting ONNX node with type "PRelu"    
> DEBUG burn_import::onnx::from_onnx: renaming node "/relu1/PRelu"    
>  WARN burn_import::onnx::dim_inference: Must implement dimension inference for PRelu    
> DEBUG burn_import::onnx::from_onnx: checking inputs for node "prelu1"    
> DEBUG burn_import::onnx::from_onnx: 
> ```

Yeah, you will need to add dim inference in dim_inference.rs file. You can use same as input function since the output will have the same dimensions.

**Comment by antimora** [comment]
Also please update the supported https://github.com/tracel-ai/burn/blob/main/crates/burn-import/SUPPORTED-ONNX-OPS.md as well

**Comment by Arjun31415** [comment]
Ok so I think what is pending is the `dim_inference` and loading the `weight` from the onnx into `alpha` of the Burn `PreluConfig`. I need help with this, where exactly do I do the latter part

**Comment by Arjun31415** [comment]
if I do those two then tests would also be fixed

**Comment by Arjun31415** [comment]
Ok I think dim inference is also fixed. Now I just need to load the prelu weight into burn

**Comment by Arjun31415** [comment]
Ok I think it is done

**Review by antimora — APPROVED** [review]
Looks great!

Thank you for your contribution to the Burn project!

**Pull request merged**